File: wearapp/src/main/java/com/thewizrd/simpleweather/preferences/OSSCreditsPreference.java
package com.thewizrd.simpleweather.preferences;

import android.content.Context;
import android.preference.Preference;
import android.util.AttributeSet;
import android.view.View;
import android.widget.TextView;

import com.thewizrd.shared_resources.AsyncTask;
import com.thewizrd.simpleweather.R;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.concurrent.Callable;

public class OSSCreditsPreference extends Preference {

    public OSSCreditsPreference(Context context) {
        super(context);
    }

    public OSSCreditsPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public OSSCreditsPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public OSSCreditsPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    protected void onBindView(View view) {
        super.onBindView(view);

        TextView textView = view.findViewById(R.id.textview);

        // Load html to string
        String creditsText = new AsyncTask<String>().await(new Callable<String>() {
            @Override
            public String call() {
                StringBuilder sBuilder = new StringBuilder();
                try (InputStreamReader sReader = new InputStreamReader(getContext().getAssets().open("credits/licenses.txt"))) {

                    int c = 0;
                    while ((c = sReader.read()) != -1) {
                        sBuilder.append((char) c);
                    }
                } catch (IOException e) {
                    //e.printStackTrace();
                }

                return sBuilder.toString();
            }
        });

        textView.setText(creditsText);
    }
}

File: wearapp/src/main/java/com/thewizrd/simpleweather/preferences/OSSCreditsPreference.kt
package com.thewizrd.simpleweather.preferences

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.TextView
import androidx.preference.Preference
import androidx.preference.PreferenceViewHolder
import com.thewizrd.shared_resources.appLib
import com.thewizrd.simpleweather.R
import kotlinx.coroutines.*
import java.io.InputStreamReader

class OSSCreditsPreference : Preference {
    private var loadJob: Job? = null

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes)

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)

        val textView = holder.findViewById(R.id.textview) as TextView
        val progressBar = holder.findViewById(R.id.progressBar)

        loadJob?.cancel()
        loadJob = appLib.appScope.launch(Dispatchers.Main.immediate) {
            supervisorScope {
                progressBar.visibility = View.VISIBLE

                val creditsText = withContext(Dispatchers.IO) {
                    val sBuilder = StringBuilder()

                    InputStreamReader(context.assets.open("credits/licenses.txt")).use { sReader ->
                        var c: Int
                        while (sReader.read().also { c = it } != -1) {
                            sBuilder.append(c.toChar())
                        }
                    }

                    sBuilder.toString()
                }

                ensureActive()

                textView.text = creditsText
                progressBar.visibility = View.GONE
            }
        }
    }

    override fun onPrepareForRemoval() {
        super.onPrepareForRemoval()
        loadJob?.cancel()
    }
}